keel-core/src/main/kotlin/com/netflix/spinnaker/keel/api/Resource.kt

```diff
@@ -31,14 +30,34 @@ data class Resource<T : ResourceSpec>(
 ) {
   init {
     require(kind.isNotEmpty()) { "resource kind must be defined" }
-    require(metadata["uid"].isValidULID()) { "resource uid must be a valid ULID" }
-    require(metadata["name"].isValidName()) { "resource name must be a valid name" }
+    require(metadata["id"].isValidId()) { "resource id must be a valid id" }
     require(metadata["serviceAccount"].isValidServiceAccount()) { "serviceAccount must be a valid service account" }
     require(metadata["application"].isValidApplication()) { "application must be a valid application" }
   }
 
   constructor(resource: SubmittedResource<T>, metadata: Map<String, Any?>) :
     this(resource.apiVersion, resource.kind, metadata, resource.spec)
+
+  // TODO: this is kinda dirty, but because we add uid to the metadata when persisting we don't really want to consider it in equality checks
+  override fun equals(other: Any?): Boolean {
+    if (this === other) return true
+    if (javaClass != other?.javaClass) return false
+
+    other as Resource<*>
+
+    if (apiVersion != other.apiVersion) return false
+    if (kind != other.kind) return false
+    if (spec != other.spec) return false
+
+    return true
+  }
```

Comment:
I'm not real happy with this, but several tests check for equality of a `Resource` (or an `Environment` or `DeliveryConfig` containing resources) before and after persistence and the fact that the "uid" metadata field is assigned at persistence time means that breaks. Open to any suggestions on improving that.